### PR TITLE
FiSHLiM: Support for Diffie-Hellmann Key-Exchange + more commands

### LINKED ISSUE #1083: FiSHLiM keyx to exchange keys with another user
The current FiSHLiM plugin does not support any key exchange between clients.
This is working in xchat old version but it is not compatible with hexchat.

When trying to do /keyx you get keyx :Unknown command
When another client tryes to do /keyx ,  you get this (random example taken from IRC):
DH1080_INIT_cbc hi4f7k4P3JF5v4k8jVcHs8WLfk8WKUWb6HDR+k0YKc7cZHp6zfAoM5EZi8dw0abpkq6keTame9S+aLk6tktqLtDSdWCBdKM0/C/w5PeAxXlNr6hORlG+jf/Uy7ICNxaXpxCvtTuORtcbmz4iQqAA3KLplaeA/t5hQOWCEHMvSfxLBvgqfMIHA

I believe that being able to do key exchange with other clients is essential part of fishlim encryption and whitout it is not compatible with other clients running different types of FiSHLiM plugins, also wraith botnet that uses random keys to encrypt the communication and only way to get the key is accepting the key from the client, in this case wraith bot.

### PR DESCRIPTION
**Pull request by philicious**
I was a long time XChat+XFiSH user but moved away due to lack of TLS support. Happily found HexChat as I always liked the clean design and scriptability/extensibility. 
Unfortunately the FiSHLiM plugin lacked essential support for key-exchange. I needed it, so I added it :)

Features added:
- Diffie-Hellman Key-Exchange (DH class taken from the guys over at https://github.com/hdznrrd/mircryption-xchat2 as its clean and uses "openssl/dh" , thx!)
- commands
  - KEYX [\<nick\>], performs DH1080 key-exchange with <nick>
  - TOPIC+ \<topic\>, sets a new encrypted topic for the current channel
  - NOTICE+ \<nick or #channel\> \<notice\>, sends an encrypted notice
  - MSG+ \<nick or #channel\> \<message\>, sends an encrypted message if you already keyx'ed

PS: the remote exploit with /keyx due to a overlong key that affected XChat/XFiSH is tested and NOT possible

fixes #1083 #653 

thx @samuellb for initial work on this plugin

PS2: I found a bug in Hexchat which has a very minor impact on this plugin  #1439
This causes if you /keyx in network window, without nick as argument, it will try to send key to \<network_name\> instead of displaying usage message

### CODE REVIEW

**Comment by TingPing** [comment]
- Random whitespaces breakages
- Don't need to reimplement base64, we already depend on multiple libraries that do it
- I'm personally not a fan of mixing in some C++ there, seems unnecessary
- Keep new features like the `/topic` change for a different PR

**Comment by philicious** [comment]
@TingPing 
- I had to use VS2013 Express (as doc suggested) where I couldnt install my AStyle addin for style-fix
- Also not a fan of C/C++ mix but I wanted to get this working asap and not re-write that entire class. also thats why I brought in its own base64 files. But you are welcomed to refactor this to your needs :)
- you can split this PR up as much as you like

**Comment by TingPing** [comment]
> But you are welcomed to refactor this to your needs :)
> you can split this PR up as much as you like

I do appreciate the work, but dropping code on a projects lap and saying "you can finish it" is not usually now it works =)

**Comment by samuellb** [comment]
Nice :) I agree with the points from TingPing, though. If you want this included in FiSHLiM you must least add MIT license headers, like in the other FiSHLiM files.

**Comment by philicious** [comment]
@samuellb I dont need my code to get integrated into official Hexchat. I am happy with it in my fork and I can build it anymtime. I'll put up binaries for others so you this doesnt need to be merged

**Comment by TingPing** [comment]
> @samuellb I dont need my code to get integrated into official Hexchat. I am happy with it in my fork and I can build it anymtime.

Then why open a PR if you have no interest in this...

> Nice :) I agree with the points from TingPing, though. If you want this included in FiSHLiM you must least add MIT license headers, like in the other FiSHLiM files.

@samuellb Feel free to finish up the work on this and open another PR, otherwise this might as well be closed.

**Comment by samuellb** [comment]
@ philicious Are you ok if we license the files under MIT license (like the other files)? And add © 2015 <the.cypher@gmail.com> to the changed files of course.

@TingPing Not sure if I have time to work on this. But I can create an improved PR for the /topic+ /msg+ /notice+ commands.

**Comment by philicious** [comment]
@TingPing I opened the PR because others already expressed interest in this in two issues. But please understand that I dont have much time to put more work into this. Of course I do understand your project "policies"

@samuellb of course I'm fine with it. As long as you mention my mail with this work you can re-license/refactor as much as you like :) thx !

**Comment by samuellb** [comment]
@philicious Thanks :)

**Comment by TingPing** [comment]
>  But please understand that I dont have much time to put more work into this.

> @TingPing Not sure if I have time to work on this.

No worries.


Just for future reference:

> This causes if you /keyx in network window, without nick as argument, it will try to send key to <network_name> instead of displaying usage message

The proper way to handle that is not relying on names, every context has a 'type' and it knows if its a channel or not. See hexchat_list_get();

**Comment by jwgn** [comment]
Could you create an x64 build of this plugin? This one does not seem to work on HexChat x64. Would be great, thank you.

**Comment by philicious** [comment]
@jwgn ye. In fact I do have/use a fixed version with x64 support but I didnt care about pushing the fixes as HexChat project didnt show much interest in this PR. I'll send you a x64-compatible version soon

**Comment by TingPing** [comment]
> I didnt care about pushing the fixes as HexChat project didnt show much interest in this PR. 

To be fair I did a basic review on what I thought needed to be done and would still like it to be done.

**Comment by philicious** [comment]
@TingPing ye ok. I have several weeks of holidays upcoming and will allot some time to fixup this PR in respect of your comments

**Comment by jwgn** [comment]
@philicious hey you said you're about to send me an x64-compatible version of your plugin soon. still waiting :)

**Comment by hxe** [comment]
you should make your own repo with just the plugin. x86 and x64

**Comment by philicious** [comment]
@jwgn sorry, for the delay, was busy .. here is the x64 version of the plugin. Confirmed to work with x64 HexChat https://mega.nz/#!mthHCI5b!ZJlnt9IoYv-P4gu7iTICjLIUrvfrhbTvgdp7qeHiEe0

@hxe I do have a forked repo over here https://github.com/philicious/hexchat/commits/master
I also added a release there with the binaries of the plugin https://github.com/philicious/hexchat/releases/tag/fishlim_v0.1.0

**Comment by jwgn** [comment]
@philicious Thank you! Had to install this runtime first to get it working:
https://www.microsoft.com/de-de/download/details.aspx?id=48145

Now keyx is working. Perfect :)

**Comment by philicious** [comment]
fixed up PR to build on Linux, here is the build, https://github.com/philicious/hexchat/releases/download/fishlim_v0.1.0/fishlim_Ubuntu14_trusty.rar

I didnt test it, but MacOSX should also build now.

**Comment by TingPing** [comment]
So I did a quick pass over the current branch and there are still far too many issues:

- Numerous leaks
- Incorrectly asserts '#' is used as a channel prefix
- Your `stricmp()` change should just use glib
- Numerous const warnings
- Comparing pointers against 0 instead of NULL

This is on top of the original comments that were not fixed and I have yet to actually look at the logic considering everything else.

**Comment by philicious** [comment]
thx for feedback @TingPing , appreciated

this PR is kinda WIP now.

I meanwhile discovered /keyx being broken on Linux (actual decryption works if you /setkey ..)
However I fear I dont have the time atm to debug/fix this plugin up for also Mac/Linux. But I still wanna help out.
Maybe @samuellb could join again and help ?

**Comment by philicious** [comment]
@TingPing , user @Groil fixed up things you mentioned. Also `/me` gets encrypted now also. Maybe you wanna review the code again if its mergable now or what would still need to be refactored. Thx

**Comment by TingPing** [comment]
I would still strongly prefer the C++ file be rewritten in C, it isn't like it actually makes use of any part of the language.

**Comment by TingPing** [comment]
Whitespace is incorrect throughout.

File: plugins/fishlim/plugin_hexchat.c [hunk 1]
```diff
@@ -239,16 +338,134 @@ static int handle_delkey(char *word[], char *word_eol[], void *userdata) {
...
+	char* topic = word_eol[2];
+
+	if (topic == 0 || *topic == 0)
+	{
+		hexchat_printf(ph, usage_topic);
+		return HEXCHAT_EAT_ALL;
+	}
+
+	target = (char *)hexchat_get_info(ph, "channel");
+	if (target == NULL || (*target != '#' && *target != '&'))
```
Comment (TingPing) [inline plugins/fishlim/plugin_hexchat.c]: Same chantype comment.

**Comment by TingPing** [comment]
Ok so I cleaned up and rebased most of the commits. There are a few remaining issues I need to iron out and I need to redo the commits that fix escaping the nick. Would be nice if any of you could test it.

**Comment by philicious** [comment]
@TingPing awesome ! If you would provide precompiled binaries (x86/x64) for the plugin, I could test it. (I dont have a Hexchat build setup anymore, so I cannot quickly build it myself. )
I'm using this on a daily basis so I'm happy to test anything around it.

**Comment by TingPing** [comment]
@tomek ^

**Comment by tomek** [comment]
@philicious https://dl.hexchat.net/hexchat/testing/tomek/fishlim/

**Comment by philicious** [comment]
@tomek thx, but your link leads to the Hexchat installer, not the fishlim plugin dlls

**Comment by tomek** [comment]
well... just install it over your current HC, you won't loose anything. If your really want just the dlls, I could upload them later (after work - so around 17-18 CEST)

**Comment by philicious** [comment]
@tomek I'm using Hexchat in portable mode. I know I could still install over it again but I'd prefer standalone plugin dlls for now, during testing. 
No hurries. just lemme know when you got to it then

**Comment by TingPing** [comment]
> I'm using Hexchat in portable mode. I know I could still install over it again but I'd prefer standalone plugin dlls for now, during testing. 

FYI, don't do that. Just install into a new directory and copy whatever over.

**Comment by philicious** [comment]
@tomek @TingPing I did a test now with /keyx /topic+ /msg+ /notice+ and /me and its working as expected :+1: 
(if it matters: I did install to new dir, portable, copied over my "config" dir from other install)

I guess you wanna bump the version here https://github.com/hexchat/hexchat/blob/wip/fishlim/plugins/fishlim/plugin_hexchat.c#L42 
in my fork I did use 0.1.0 so you may wanna use sth higher then.

**Comment by TingPing** [comment]
@philicious Curious what other clients, if any, you tested talking with? Yea I'll bump the ver before merge.

**Comment by philicious** [comment]
@TingPing I tested with a `mirc v7.46` and fish from https://syndicode.org/fish_10/

**Comment by TingPing** [comment]
I had a brief test with a Textual user and the exchange seemed to work fine and they could read my text but I couldn't read theirs. I need to test it further but I suspect this was always an issue unrelated to keyx.

**Comment by philicious** [comment]
@TingPing if you wanna make sure its not /keyx related, you and your chat partner do a `/setkey` with a key both agreed on, then test if you can read each other

(Also you could try with my plugin to see if it was working before refactoring)

**Comment by TingPing** [comment]
Planned on it, just happens I am awake when they are asleep :P

**Comment by philicious** [comment]
@TingPing any luck ?

**Comment by TingPing** [comment]
Merged.

**Comment by idnow-philicious** [comment]
@TingPing for what its worth: I tested the plugin on MacOS now. HexChat installed via homebrew. The added features are working like a charm.

**Comment by philicious** [comment]
@TingPing for what its worth: I tested the plugin on MacOS now. HexChat installed via homebrew. The added features are working like a charm.

**Pull request closed without merging**